### DISCUSSION TITLE
How to add SHX fonts for use by Matplotlib drawing back-end

### DISCUSSION THREAD

**Discussion by IvanKachaikinCendas** [mozman/ezdxf#710] [Q&A] [answered]
Hi @mozman! First of all, thanks for an awesome package!🤩

While running the rendering using `ezdxf`, I often get a warning message saying something like:

`findfont: Font family ['isocp'] not found. Falling back to DejaVu Sans.`

I succeeded to fix an issue for the fonts available as TFF, however still getting the same warning for the AutoCAD fonts saved as SHX. So my question is: is there a way to somehow import the SHX fonts from AutoCAD into `ezdxf`?

Thanks in advance for your feedback!🙂

**Comment by mozman** [comment]
The SHX font format is not documented or supported by many libraries/packages like `Matplotlib` and `qt`, therefore only SHX fonts which have corresponding TTF-fonts can be rendered by these backends.

Sometimes the `Matplotlib` package is stubborn and does not rebuild the font-cache properly if new fonts are installed, deleting the `fontlist-v???.json` file in `~/.matplotlib` may help in this case.

`Ezdxf` uses it's own font-cache for faster font-measurement if you use the `Matplotlib`  font support which is the default case.
If you use special fonts which are not included in the default cache files of `ezdxf`, you have to rebuild the font-cache: https://ezdxf.mozman.at/docs/options.html#font-cache-directory

... but "isocp" is included in the default cache files

Edit: usual the "isocp" font should be mapped to "isocp___.ttf", maybe you can upload a simple example DXF which shows your issue.

**Comment by IvanKachaikinCendas** [comment]
Hi @mozman! Ok got it, thanks for your feedback! Will retry with `fontlist-v???.json` delete in `~/.matplotlib`.

Regarding the "isocp" issue, I created a dummy DXF with some text using "Txt" and "isocp" fonts. By default, it returns me this warning for both of these fonts and draws the text using some default font. You can find this dummy DXF below:
[fonts.dxf.zip](https://github.com/mozman/ezdxf/files/9165688/fonts.dxf.zip)

Please note: I run the `ezdxf==0.17.1` and not the later versions for some backward compatibility reasons. I also run it on MacOS that may be another reason of such an issue.

**Comment by mozman** [comment]
I tested your `fonts.dxf`: 

- Windows with an Autodesk product (TrueView) installed - no message
- Linux without an Autodesk product: `font ... not found message`

Next I copied on Linux the replacement fonts into `~/.fonts`, same message `font ... not found message`

Deleted `fontlist-v330.json` from  `~/.cache/matplotlib` and everything worked

![Screenshot from 2022-07-22 10-34-39](https://user-images.githubusercontent.com/4230966/180399266-bb1bdeaa-984c-4303-b454-dedefe00a8a6.png)

Conclusion: the TTF replacement fonts are not available system wide, or Matplotlib can't find it, or the Matplotlib font cache is outdated.

**Comment by IvanKachaikinCendas** [comment, marked as answer]
@mozman Ok sounds clear, thanks for your feedback!🙂

By the way, I also found some solution that is a bit ugly from the software engineering and the code readability point of view, but that does the trick in case when someone has a folder with .ttf files that they don't want to install globally (like it was in my case). More specifically, if `custom_fonts_paths` is a `list` with paths to the folders with custom fonts, then the following worked for me:

```python
from os.path import join
import matplotlib.font_manager as font_manager

for font in font_manager.findSystemFonts(custom_fonts_paths):
    font_manager.fontManager.addfont(font)


class FixedFontManager(font_manager.FontManager):
    def __init__(self, size=None, weight='normal'):
        super().__init__(size, weight)
        for font in font_manager.findSystemFonts(custom_fonts_paths):
            self.addfont(font)


font_manager.FontManager = FixedFontManager
```

Note: the code above must be placed **before** the code that calls the `ezdxf` export to PDF/JPG/etc. **in the same file**.
